lib/mix/test/mix/tasks/escript_test.exs

```diff
@@ -195,4 +195,106 @@ defmodule Mix.Tasks.EscriptTest do
+        end
+      end
+      """
+
+      System.cmd("git", ~w[add .])
+      System.cmd("git", ~w[commit -m "ok"])
+
+      send self, {:mix_shell_input, :yes?, true}
+      Mix.Tasks.Escript.Install.run ["git", File.cwd!()]
+      assert_received {:mix_shell, :info, ["Generated escript git_repo with MIX_ENV=prod"]}
+
+      escript_path = Path.join([tmp_path(".mix"), "escripts", "git_repo"])
+      assert System.cmd("escript", [escript_path]) == {"TEST\n", 0}
+    end
+  after
+    purge [GitRepo, GitRepo.Mixfile]
+  end
+
+  defmodule TestFetch do
+    def fetch(package_name, dep_spec, _in_package) do
+      Agent.update(__MODULE__, fn _ -> {package_name, dep_spec} end)
+    end
+  end
+
+  test "escript.install git" do
+    assert_installed ["git", "https://example.com/user/repo.git"],
+      {"new escript", {:"new escript", [branch: "master", git: "https://example.com/user/repo.git", submodules: nil]}}
+  end
+
+  test "escript.install git branch" do
+    assert_installed ["git", "https://example.com/user/repo.git", "branch", "not_master"],
+      {"new escript", {:"new escript", [branch: "not_master", git: "https://example.com/user/repo.git", submodules: nil]}}
+  end
+
+  test "escript.install git ref" do
+    assert_installed ["git", "https://example.com/user/repo.git", "ref", "not_master"],
+      {"new escript", {:"new escript", [ref: "not_master", git: "https://example.com/user/repo.git", submodules: nil]}}
+  end
+
+  test "escript.install git tag" do
+    assert_installed ["git", "https://example.com/user/repo.git", "tag", "not_master"],
+      {"new escript", {:"new escript", [tag: "not_master", git: "https://example.com/user/repo.git", submodules: nil]}}
+  end
+
+  test "escript.install git submodules" do
+    assert_installed ["--submodules", "git", "https://example.com/user/repo.git"],
+      {"new escript", {:"new escript", [branch: "master", git: "https://example.com/user/repo.git", submodules: true]}}
+  end
+
+  test "escript.install git app" do
+    assert_installed ["--app", "my_app", "git", "https://example.com/user/repo.git"],
+      {"my_app", {:my_app, [branch: "master", git: "https://example.com/user/repo.git", submodules: nil]}}
+  end
+
+  test "escript.install github" do
+    assert_installed ["github", "user/repo"],
+      {"new escript", {:"new escript", [branch: "master", git: "https://github.com/user/repo.git", submodules: nil]}}
+  end
+
+  test "escript.install hex" do
+    assert_installed ["hex", "a_package"],
+      {"a_package", {:a_package, ">= 0.0.0", [hex: :a_package]}}
+  end
+
+  test "escript.install hex app" do
+    assert_installed ["--app", "app_name", "hex", "a_package"],
+      {"app_name", {:app_name, ">= 0.0.0", [hex: :a_package]}}
+  end
+
+  test "escript.install hex version spec" do
+    assert_installed ["hex", "a_package", "1.0.0"],
+      {"a_package", {:a_package, "1.0.0", [hex: :a_package]}}
+  end
+
+  defp assert_installed(argv, expected_agent_state) do
```

Comment:
I think there is a better way to test this. Instead of relying on callbacks, have a function in Mix.Local.Installer (or Mix.Local.Fetcher) that will parse the arguments and return exactly how it should be handled. For example:

    parse_args("foo/bar")
    #=> {:local, "foo/bar"}

    parse_args("http://foo.org/bar")
    #=> {:url, "foo/bar"}

    parse_args("hex phoenix 0.3.0")
    #=> {:fetcher, {:phoenix, "0.3.0"}}

    parse_args("github elixir-lang/plug")
    #=> {:fetcher, {:plug, github: "elixir-lang/plug"}} 

This way, you will be able to test everything without really needing to pass callbacks. The integration tests for local, URL and fetcher will be able to handle all cases.